File: internal/scraper/creators/polyhaven.go
package scraper

import (
	"context"
	"fmt"
	"log/slog"

	"github.com/struffel/3d-assets-one/internal/model"
)

type PolyhavenScraper struct{}

func (s *PolyhavenScraper) Creator() model.Creator { return model.CreatorPolyhaven }

var phTypeMapping = map[int]model.AssetType{
	0: model.AssetTypeHDRI,
	1: model.AssetTypePBRMaterial,
	2: model.AssetTypeModel3D,
}

func (s *PolyhavenScraper) ScrapeAssets(ctx context.Context, existing []model.StoredAsset) ([]ScrapedAsset, error) {
	const maxPerRun = 10
	const apiURL = "https://api.polyhaven.com/assets"
	const viewBase = "https://polyhaven.com/a/"
	const thumbPrefix = "https://cdn.polyhaven.com/asset_img/thumbs/"
	const thumbSuffix = ".png?height=512"

	var resp map[string]struct {
		Name          string   `json:"name"`
		Tags          []string `json:"tags"`
		Type          int      `json:"type"`
		DatePublished int64    `json:"date_published"`
	}

	if _, err := FetchJSON(ctx, apiURL, &resp); err != nil {
		return nil, fmt.Errorf("polyhaven api: %w", err)
	}

	var results []ScrapedAsset
	for key, ph := range resp {
		if len(results) >= maxPerRun {
			break
		}

		url := viewBase + key
		if ContainsURL(existing, url) {
			continue
		}

		img, err := FetchImage(ctx, thumbPrefix+key+thumbSuffix)
		if err != nil {
			slog.Warn("polyhaven: thumbnail fetch failed", "key", key, "err", err)
			continue
		}

		at := phTypeMapping[ph.Type]
		results = append(results, ScrapedAsset{
			Title:        ph.Name,
			URL:          url,
			Tags:         ph.Tags,
			Type:         at,
			Creator:      model.CreatorPolyhaven,
			Status:       model.ScrapedNewlyFound,
			RawThumbnail: img,
		})
	}

	return results, nil
}

func (s *PolyhavenScraper) ValidateAsset(ctx context.Context, asset model.StoredAsset) bool {
	return HeadOK(ctx, asset.URL)
}

File: app/creator/logic/polyhaven.py
from app.asset.asset_type import AssetType
from app.asset.scraped_asset import ScrapedAsset
from app.asset.scraped_asset_collection import ScrapedAssetCollection
from app.asset.scraped_asset_status import ScrapedAssetStatus
from app.asset.stored_asset_collection import StoredAssetCollection
from app.creator.creator import Creator
from app.creator.creator_logic import CreatorLogic
from app.fetch.web_item_reference import WebItemReference


class CreatorLogicPolyhaven(CreatorLogic):
    creator = Creator.POLYHAVEN
    api_url = "https://api.polyhaven.com/assets"
    view_base_url = "https://polyhaven.com/a/"
    thumbnail_url_prefix = "https://cdn.polyhaven.com/asset_img/thumbs/"
    thumbnail_url_suffix = ".png?height=512"
    type_mapping = {0: AssetType.HDRI, 1: AssetType.PBR_MATERIAL, 2: AssetType.MODEL_3D}
    max_assets_per_run = 10

    def scrape_assets(self, existing_assets: StoredAssetCollection) -> ScrapedAssetCollection:
        tmp_collection = ScrapedAssetCollection()
        result = WebItemReference(self.api_url).fetch().parse_as_json()
        if result is None:
            raise RuntimeError(f"Could not fetch or parse Polyhaven API data from {self.api_url}")

        for key, ph_asset in result.items():
            url = self.view_base_url + key
            if not existing_assets.contains_url(url) and len(tmp_collection) < self.max_assets_per_run:
                thumbnail_url = self.thumbnail_url_prefix + key + self.thumbnail_url_suffix
                raw_thumbnail = WebItemReference(url=thumbnail_url).fetch().parse_as_image()

                asset = ScrapedAsset(
                    id=None,
                    creator_given_id=None,
                    url=url,
                    title=ph_asset.get("name", key),
                    tags=ph_asset.get("tags", []),
                    type=self.type_mapping.get(int(ph_asset.get("type", 0)), AssetType.OTHER),
                    creator=Creator.POLYHAVEN,
                    status=ScrapedAssetStatus.NEWLY_FOUND,
                    raw_thumbnail=raw_thumbnail,
                )
                tmp_collection.append(asset)

        return tmp_collection
